rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,263 @@
+
+[AWS](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNS-failover.html) (well understood)
+[Google DNS](https://cloud.google.com/DNS/docs/zones/manage-routing-policies#before_you_begin)
+[Azure](https://learn.microsoft.com/en-us/azure/traffic-manager/traffic-manager-monitoring)
+
+## OCM / ACM
+This is very similar, with the caveat that the "cluster" is the hub cluster for this OCM and that the gateway will be a Kuadrant gatewayClass. You can think about this setup as "single cluster" with a gateway that has multiple addresses. It will continue to work with these changes. 
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+## General Logic Overview
+
+### Kuadrant Operator
+
+The general flow in the Kuadrant operator follows a single path, where it will always output a valid DNSRecord, which specifies everything needed for the listener host of a given gateway on the cluster, and is unaware of the concept of distributed DNS. The core logic for doing this already exists.
+
+
+### DNS Operator / controller
+
+We will update the DNS Operator to leverage the external DNS provider logic (IE effectively use it as a library) and layer our multi-cluster changes ontop of this logic. Our hope is to keep what we are doing as compatible with external DNS as possible so that we can contribute / propose changes to external-dsn
+The DNSRecord created by the Kuadrant operator is reconciled by the DNSRecord controller as part of the DNS Operator component (the operator and controller terms can be used interchangeably). In the event of a change to the endpoints in the DNSRecord resource (including deletion and creation of the resource), the DNSRecord controller will first pull down the relevant records for that DNS name from the provider and store them in the DNSRecord status. Next it will update the affected endpoints in that record set and validate the record set has no "dead ends" before writing the changes back to the DNS Provider's zone. Once the remote write is successful, it will then re-queue the DNSRecord for validation. The validation will be the same validation as previously done before the write. It will pull the zone and ensure that the endpoints from its own local DSNRecord are reflected correctly. If the validation is successful, the controller will mark this in the status of the DNSRecord and will re-queue the validation for ~15 (example) minutes later (a stable state verification). If validation is unsuccessful, it will re-queue the DNSRecord for validation rapidly (5 seconds for example). With each re-queue after an unsuccessful validation, it will add a random amount of "jitter" time to increase the chance it moves out of sync with any other actors. Each time it re-queues, it will mark this in the status (see below). At any point, if there is a change to the on-cluster DNSRecord, this back off and validation will be reset and started again.  
```

Comment:
What's the thinking behind adding the list of zone endpoints to the status? Is that just for visibility i.e. this list should eventually be the same on all DNSRecord instances giving you a full view of the world on any cluster?

Reply:
Yes I think it is useful to be able to see what is the state that the controller is working from and allow comparision with controllers in other clusters